analyzers/src/SonarAnalyzer.CSharp/Rules/LoggingArgumentsShouldBePassedCorrectly.cs

```diff
@@ -123,4 +124,7 @@ private static void CheckInvalidTypeParams(InvocationExpressionSyntax invocation
 
     private static bool IsInvalidArgument(ArgumentSyntax argumentSyntax, SemanticModel model, ImmutableArray<KnownType> knownTypes) =>
         model.GetTypeInfo(argumentSyntax.Expression).Type?.DerivesFromAny(knownTypes) is true;
+
+    private static ImmutableArray<KnownType> Filter(IMethodSymbol methodSymbol, ImmutableArray<KnownType> knownTypes) =>
+        knownTypes.Where(knownType => !methodSymbol.ConstructedFrom.Parameters.Any(x => x.Type.DerivesFrom(knownType))).ToImmutableArray();
```

Comment:
Why `ConstructedFrom`?
If the generic in the _constructed_ type is, say, `Exception`, doesn't that mean that the user passed it in the correct spot?

For example on this method:
```csharp
{NLog.Logger.ConditionalTrace<System.Exception>(string, System.Exception)}
```

you consider this `Noncompliant`:

```csharp
logger.ConditionalTrace("Message!", exception);
```

...even though here `ConditionalTrace<T>` is constucted as `ConditionalTrace<Exception>`

Reply:
I've intentionally used `ConstructedFrom` exactly for this :). What happens with type arguments is that they will be grouped together and passed forward as `params object[]`. See: https://github.com/NLog/NLog/blob/dev/src/NLog/Logger-generated.cs#L515

So we want to consider them as noncompliant.

Reply:
Good to know, thanks!
Resolving.